include/anvill/Arch.h

```diff
@@ -0,0 +1,186 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include <llvm/IR/DebugInfoMetadata.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/IntrinsicInst.h>
+#include <llvm/IR/Metadata.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Type.h>
+
+#include "Decl.h"
+
+namespace remill {
+class Arch;
+class IntrinsicTable;
+struct Register;
+}  // namespace remill
+
+namespace anvill {
+
+enum SizeConstraint : unsigned {
+  kBit8 = (1 << 0),
+  kBit16 = (1 << 1),
+  kBit32 = (1 << 2),
+  kBit64 = (1 << 3),
+  kBit128 = (1 << 4),
+
+  kMaxBit128 = kBit128 | kBit64 | kBit32 | kBit16 | kBit8,
+  kMaxBit64 = kBit64 | kBit32 | kBit16 | kBit8,
+  kMaxBit32 = kBit32 | kBit16 | kBit8,
+
+  kMinBit128 = kBit128,
+  kMinBit64 = kBit128 | kBit64,
+  kMinBit32 = kBit128 | kBit64 | kBit32,
+};
+
+enum TypeConstraint : unsigned {
+  kTypeInt = (1 << 0),
+  kTypePtr = (1 << 1),
+  kTypeFloat = (1 << 2),
+  kTypeVec = (1 << 3),
+
+  kTypeIntegral = kTypeInt | kTypePtr,
+};
+
+struct VariantConstraint {
+  VariantConstraint(std::string _register_name, TypeConstraint _type_constraint,
+                    SizeConstraint _size_constraint)
+      : register_name(_register_name),
+        type_constraint(_type_constraint),
+        size_constraint(_size_constraint) {}
+
+  std::string register_name;
+  TypeConstraint type_constraint;
+  SizeConstraint size_constraint;
+};
+
+// Captures the different sizes and constraints of a single register
+struct RegisterConstraint {
+  RegisterConstraint(std::vector<VariantConstraint> _variants)
+      : variants(_variants) {}
+
+  std::vector<VariantConstraint> variants;
+};
+
+class CallingConvention {
+ public:
+  CallingConvention(llvm::CallingConv::ID _identity) : identity(_identity) {}
+  virtual ~CallingConvention(){};
+
+  virtual std::vector<ParameterDecl> BindParameters(
+      const llvm::Function &function) = 0;
+  virtual std::vector<ValueDecl> BindReturnValues(
+      const llvm::Function &function) = 0;
+  virtual remill::Register *BindReturnStackPointer(
+      const llvm::Function &function) = 0;
+
+  llvm::CallingConv::ID getIdentity() { return identity; }
+
+ private:
+  llvm::CallingConv::ID identity;
+};
+
+class X86_64_SysV : public CallingConvention {
+ public:
+  X86_64_SysV() : CallingConvention(llvm::CallingConv::X86_64_SysV) {}
+  ~X86_64_SysV(){};
+  std::vector<ParameterDecl> BindParameters(const llvm::Function &function);
+  std::vector<ValueDecl> BindReturnValues(const llvm::Function &function);
+  remill::Register *BindReturnStackPointer(const llvm::Function &function);
+
+ private:
+  // TODO: add the rest of the Registers here, not just the 32-bit and 64-bit
```

Comment:
All `TODO` comments should have a username in them, e.g. `TODO(aaronyoo): blah blah` so then someone looking at the code knows who to ask about the context of that TODO.